trl/trainer/grpo_trainer.py

```diff
@@ -1431,8 +1475,155 @@ def _generate(self, prompts: list):
+            idxs_with_tool = [idx for idx, tool_call in enumerate(tool_calls) if tool_call]
+            tool_calls = [tool_calls[idx] for idx in idxs_with_tool]
+            tool_mask = [[0] * len(ids) for ids in completion_ids]
+            tool_call_count = 0
+            tool_failure_count = 0
+        else:
+            idxs_with_tool = []
+            tool_mask = None
+
+        # Tool execution loop: execute tools, then regenerate completions with tool results appended to the prompt
+        while idxs_with_tool:
+            prompt_completion_tools = [prompts[i] for i in idxs_with_tool]  # select only prompts that need tool calls
+
+            # Tokenize the current prompt. We will use this to filter out overlong samples later.
+            kwargs = {
+                "tools": self.tools,
+                "add_generation_prompt": True,
+                "tokenize": True,
+                "chat_template": self.chat_template,
+                **self.chat_template_kwargs,
+            }
+            p_ids = self.processing_class.apply_chat_template(prompt_completion_tools, **kwargs)["input_ids"]
+
+            # Call the tools, and build the new prompt for generation
+            for idx in range(len(idxs_with_tool)):
+                idx_with_tool = idxs_with_tool[idx]
+                tool_call_list = tool_calls[idx]
+                prompt_completion_tool = prompt_completion_tools[idx]
+                prompt_completion_tool.append(completions[idx_with_tool][-1])
+                for tool_call in tool_call_list:
+                    tool_call_count += 1
+                    if tool_call["type"] == "function":
+                        function = tool_call["function"]
+                        name = function["name"]
+                        try:
+                            result = self._tool_dict[name](**function["arguments"])
+                        except Exception as e:
+                            tool_failure_count += 1
+                            result = {"error": str(e)}
+                    else:
+                        tool_failure_count += 1
+                        name = tool_call.get("name", "unknown")
+                        result = {"error": f"Unsupported tool call type: {tool_call['type']}"}
+                    tool_message = {"role": "tool", "name": name, "content": str(result)}
+                    prompt_completion_tool.append(tool_message)
+                    completions[idx_with_tool].append(tool_message)
+
+            # Tokenize and filter samples whose length exceeds max allowed length. This is important, because if vLLM
+            # is called with an input longer than its max model length, it will error out.
+            pct_ids = self.processing_class.apply_chat_template(prompt_completion_tools, **kwargs)["input_ids"]
+            overlong = [len(pct) - len(p) >= self.max_completion_length for p, pct in zip(p_ids, pct_ids, strict=True)]
```

Comment:
Should this refer instead to `max_model_len` because technically one could have a small `max_completion_length` but a long context that is built up over many steps of turn calling.